httpbakery/client.go

```diff
@@ -261,18 +261,25 @@ func (c *Client) doWithBody(req *http.Request, body io.ReadSeeker) (*http.Respon
 	if err != nil {
 		return nil, errgo.Mask(err, errgo.Any)
 	}
+	var cookiePath string
 	cookieURL := req.URL
 	if path := resp.Info.MacaroonPath; path != "" {
 		relURL, err := parseURLPath(path)
 		if err != nil {
 			logger.Warningf("ignoring invalid path in discharge-required response: %v", err)
 		} else {
-			cookieURL = req.URL.ResolveReference(relURL)
+			cookieURL := req.URL.ResolveReference(relURL)
+			cookiePath = cookieURL.Path
 		}
 	}
-	if err := SetCookie(c.Jar, cookieURL, macaroons); err != nil {
+	if len(cookiePath) > 1 && strings.HasSuffix(cookiePath, "/") {
+		cookiePath = strings.TrimSuffix(cookiePath, "/")
+	}
+	logger.Debugf("cookiePath: %q", cookiePath)
+	if err := SetCookieWithPath(c.Jar, cookieURL, cookiePath, macaroons); err != nil {
```

Comment:
I'm not entirely sure that SetCookieWithPath is justified. I think I'd just use NewCookie, set the path and use c.Jar.SetCookies.

That way we perhaps won't feel tempted to add SetCookieWithPathAndName and SetCookieWithName too.